tests/mdx/__snapshots__/jsfmt.spec.js.snap

```diff
@@ -361,16 +366,19 @@ semi: false
 </Hello>
 <Hello>
   test <World /> test
-</Hello>123
+</Hello>
+123
 
 ---
 
 <Hello>
   test <World /> test
-</Hello> 123
+</Hello>{" "}
+123
 <Hello>
   test <World /> test
-</Hello> 234
+</Hello>{" "}
```

Comment:
Not good

Reply:
Why?

Reply:
Not `prettier` to me.

`mdx` is both `html` and `jsx`.

Reply:
For example, the following is valid mdx:

```html
Inline <Test><!-- it's a comment --></Test>
```

Reply:
This specific text is JSX, not HTML. And that's how Prettier formats JSX.

Reply:
> This specific text is JSX, not HTML. And that's how Prettier formats JSX.

I have to say that's not true, see my comment above.

Reply:
Please clarify what exactly is not true. I don't understand.

Reply:
> For example, the following is valid mdx:
> And the following is not jsx:

Both your examples aren't JSX and seem to have nothing to do with what we discuss. 🤔

Reply:
I understood the text should be parsed as `jsx` now, but still think `html` style is `prettier` in this case and they are literally same, we don't need to consider it as plain `jsx`, just like our default `json` output is actually `js_expression`, right?.

Reply:
Do you have a real-life use case for text like this?

Reply:
> we don't need to consider it as plain jsx

We should consider it JSX. At least because it can contain JSX expression.

Anyway, I think it's a totally separate thing and the current PR should not take care of it. Can you open an issue?

Reply:
Yeah, this looks a bit weird, but it is how prettier formats the same code in a JSX file. It seems to make sense to format the code the same - but perhaps there are reasons to have different formatting in MDX?

Reply:
Well, although it's a bug fix PR, but it will change current behavior (whatever whether it was included as intended), so I think we should finish it up here.

Reply:
```jsx

<div/> {' '}123




<div/>

{' '}123
```

Current behavior is much better for refactoring.

Reply:
@JounQin I don't understand your last example. Please show some real-life actual code that this PR would make look significantly worse.

Reply:
@thorn0 Would the following examples make sense to you?

```jsx
// original code, current behavior
<span>1</span> 2

---

// this PR's output
<span>1</span>{' '}
2

---

// If we would to refactor it by adding two lines
<span>1</span>

{' '}
2

---

// We'll only add two lines simply on current behavior
<span>1</span>

2

```

Reply:
> Yeah, this looks a bit weird, but it is how prettier formats the same code in a JSX file. It seems to make sense to format the code the same - but perhaps there are reasons to have different formatting in MDX?

As I said, `mdx` is both `html` and `jsx`, and current behavior is much easier for refactoring as above.

Reply:
Now I see your point. But it doesn't look convincing to me. I still it's more important that people be aware that this text is JSX, which Prettier would make more clear. When those two lines are added, the text becomes markdown. To avoid mistakes, the content author must understand this clearly anyway.

Reply:
And also, `{' '}` in following examples are different:

```jsx
{' '}<span>my text</span>{' '}

prefix <span>my text</span>{' '}

<span>my text</span>{' '}
```

![image](https://user-images.githubusercontent.com/8336744/68853533-27b36f80-0715-11ea-83ac-a0a417209d82.png)


In general, inline `{' '}` is a bad idea even if it is actually `jsx`.

Reply:
You can say the same about anything in curly braces (`{ foo }`). Or in other separators, e.g. `**foo**`.

Reply:
@ikatyang What are your thoughts on this?

Reply:
I'm not hiding the fact that it's `jsx`, but `mdx` extends `markdown` and `markdown` extends `html`.

The fact is that they are just been implemented as `jsx` but they should be considered as `html` in these confusing cases. That's why the inline comments are `html` style rather than `jsx` style.

Reply:
A paragraph (thing separated by double line breaks) is either markdown (with embedded HTML) or JSX. That's it. HTML inline comments don't work inside JSX paragraphs because they make JSX invalid.

> just been implemented as jsx but they should be considered as html

That's a very wrong conclusion, I would say.

Reply:
@thorn0

```html
Inline <Test><!-- it's a comment --></Test>
```

I mean the above example, it's `jsx` + `html` style comment.

Reply:
No, it's not JSX. Come on.

Reply:
<Test> is a valid jsx element, why it's not jsx?

Reply:
Still it's not JSX. It's something else.

Reply:
So whatever `mdx` is a likely kind of *special* `jsx` like `jsx` is likely a kind of *special* `html`, we should not simply consider all of them as plain `jsx`.

I'm strongly recommending that we should not change current behavior.

Reply:
Example:
```
Inline <h3 title={ 'foo bar' }>{ 'baz' }<!-- it's a comment -->_qux_</h3> **quux**
```
It understands expressions in attributes (`{ 'foo bar' }`), but not in the inner text (`{ 'baz' }`). Also it understands markdown and HTML comments in the inner text. So this is indeed some kind of extended HTML.

But this works only in markdown paragraphs. Pure JSX paragraphs contain pure normal JSX.

Reply:
The change we discuss doesn't affect that special inline quasi-JSX. Pure JSX paragraphs where it happens are plain JSX.

**Prettier pr-6949**
[Playground link](https://deploy-preview-6949--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEATABAHlQSwG4D0AfOqiADQgQAOMO0AzsqAIYBObEA7gArsJMULPBBxlKAIzYswAazgwAytRk4oAc2Qw2AVziUAFjAC2AGwDqBnPAYqwcRQOv5rAT2TgGTSmoZw2MDzS6sYsyABmLKZ+lABWDAAeAELScgqKLMZwADJqcBFRMSDxCYpq6qZwAIo6EPAF0fogKmx+bB7GqAkUzWxqMOZiMAbIABwADJTUnH7m0tQe03BtePmUAI618EE0giAsDAC0UHBwqGc9bHCbOFdBLCFhSJGNlH7GOFq6TQzllTV1fLPQpNGAsCSDVDDZAAJko2hYOFM5QAwhBjKEPFBoGsQDo-AAVcGCF5FPB6ACSUHOsEUYD6tAAgtTFDBXJUGn4AL5coA)
```sh
--parser mdx
```

**Input:**
```mdx
d <div/> d
```

**Output:**
```mdx
d <div/> d

```

Reply:
Then the different outputs are very confusing and make refactoring bad experience as I mentioned. Current behavior doesn't change the finally output content and easier for document author.

Reply:
And I still cann't get your point that why we must change current behavior which works just fine and friendly to document authors.

Reply:
@JounQin  Uh, okay, you convinced me.

@Tigge Seems like this PR is blocked by this problem.

Reply:
@thorn0 Glad to hear that. 🍺

Then I'm going to finish my original PR based on @Tigge's great work, maybe tonight.

Reply:
I'm a little bit unsure from the discussion what is the problem. Is it the `{' '}` blocks or the lack of inline formatting? Or something else?

Reply:
@Tigge Inline text format whatever prefix or suffix, whitespace or non-whitespace.

Reply:
@Tigge The requirement seems to be as follows: in block-level JSX (paragraphs) that end with text or a JSX expression (not an element or a fragment), the line break should not be inserted between the last tag of the last element (or fragment) and the JSX node that follows it.

So implementation-wise, we need to tweak the way we print `JSXText` and `JSXExpressionContainer` nodes if they are descendants of a node marked with `__maybe_mdx_adjacent`. Might be not easy because the JSX printing code in Prettier is quite sophisticated.

Also it's not clear if we should respect `--prose-wrap`. It's ignored in JSX paragraphs now:

**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAZCAnOBbABOrPAnfTE4osyw0mi28qh6xup+j9rtn1vlgcyGdeggDpQJEgDwUAfBJH9h3ZUsHqVorWtUa9O-ds2bcuGQHp5IADQgIABxgBLaAGdkoAIYYMEAO4ACj4IHiheAG4QzgAmtiAARhheYADWcDAAyg4pzlAA5sgwGACucHYAFjDYADYA6hXO8G45YHCZoU3OEU0Ansjgbh52eW5wGDCByfnYXsgAZl41Y3YAVm4AHgBCyWkZmV7YcKh5cAtLKyDrG5l5+TVwAIolEPDny+UgORhjGAPYMQ28QcGDyMDqsRgFWQAA4AAx2EEQMZ1ZIOAZQOARcbxACOL3gU0cYRAXjcAFpMXAYtT4lh8c4sFMvDM5khFh87GNsM4iqVPm47g9nq8zuyLp8YF4EhCYlDkAAmOzFLzOGp3ADCEGwswx0DOdhKYwAKtKwhzLhEygBJKA02CZMCgpwAQTtmRgvQe7zGAF9fUA)
```sh
--parser mdx
--prose-wrap never
```

**Input:**
```mdx
Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem 


<Lorem>
Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem   
</Lorem>
```

**Output:**
```mdx
Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem

<Lorem>
  Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem
  Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem Lorem
  Lorem Lorem Lorem Lorem Lorem Lorem
</Lorem>

```

Reply:
> Also it's not clear if we should respect `--prose-wrap`.

If it does not affect the final output and user experience, we should not address it here IMO.